File: cdm/java/mil/tatrc/physiology/datamodel/properties/SEScalarEnergy.java
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

package mil.tatrc.physiology.datamodel.properties;

import com.kitware.physiology.cdm.Properties.ScalarEnergyData;

import mil.tatrc.physiology.datamodel.properties.CommonUnits.EnergyUnit;

/**
 * @author abray
 * Energy Scalar Class, will enforce that units are proper to Energys
 */

public class SEScalarEnergy extends SEScalar
{
  public SEScalarEnergy()
  {
    super();
  }
  
  /**
   * @param value
   * @param unit - specific any valid standard unit abbreviation
   * http://www.bipm.org/en/si/ for this quantity type
   */
  public SEScalarEnergy(double value, String unit)
  {
    this();
    this.setValue(value, unit);
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public SEScalarEnergy(double value, EnergyUnit unit)
  {
    this();
    this.setValue(value,unit);
  }
  
  public static void load(ScalarEnergyData src, SEScalarEnergy dst)
  {
    SEScalar.load(src.getScalarEnergy(),dst);
  }
  public static ScalarEnergyData unload(SEScalarEnergy src)
  {
    if(!src.isValid())
      return null;
    ScalarEnergyData.Builder dst = ScalarEnergyData.newBuilder();
    unload(src,dst);
    return dst.build();
  }
  protected static void unload(SEScalarEnergy src, ScalarEnergyData.Builder dst)
  {
    SEScalar.unload(src,dst.getScalarEnergyBuilder());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public void setValue(double value, EnergyUnit unit)
  {
    this.setValue(value,unit.toString());
  }
  
  /**
   * @param value
   * @param unit - enumeration of commonly used units for this type
   * Units are not limited to this set.
   * You can use the other method convention:(double,string)
   * and specify any valid standard unit abbreviation sting for this type
   * Unit Abbreviation Standards: http://www.bipm.org/en/si/
   */
  public double getValue(EnergyUnit unit)
  {
    return this.getValue(unit.toString());
  }

  public boolean validUnit(String unit)
  {
    if(EnergyUnit.validUnit(unit))
      return true;
    return false;
  }
}

File: cdm/cpp/properties/SEScalarEnergy.cpp
/* Distributed under the Apache License, Version 2.0.
   See accompanying NOTICE file for details.*/

#include "stdafx.h"
#include "properties/SEScalarEnergy.h"

const EnergyUnit EnergyUnit::J("J");
const EnergyUnit EnergyUnit::mJ("mJ");
const EnergyUnit EnergyUnit::kJ("kJ");
const EnergyUnit EnergyUnit::kcal("kcal");

bool EnergyUnit::IsValidUnit(const std::string& unit)
{
  if (J.GetString().compare(unit) == 0)
    return true;
  if (mJ.GetString().compare(unit) == 0)
    return true;
  if (kJ.GetString().compare(unit) == 0)
    return true;
  if (kcal.GetString().compare(unit) == 0)
    return true;
  return false;
}

const EnergyUnit& EnergyUnit::GetCompoundUnit(const std::string& unit)
{
  if (J.GetString().compare(unit) == 0)
    return J;
  if (mJ.GetString().compare(unit) == 0)
    return mJ;
  if (kJ.GetString().compare(unit) == 0)
    return kJ;
  if (kcal.GetString().compare(unit) == 0)
    return kcal;
  std::stringstream err;
  err << unit << " is not a valid Energy unit";
  throw CommonDataModelException(err.str());
}

void SEScalarEnergy::Load(const cdm::ScalarEnergyData& src, SEScalarEnergy& dst)
{
  SEScalarEnergy::Serialize(src, dst);
}
void SEScalarEnergy::Serialize(const cdm::ScalarEnergyData& src, SEScalarEnergy& dst)
{
  SEScalarQuantity<EnergyUnit>::Serialize(src.scalarenergy(), dst);
}

cdm::ScalarEnergyData* SEScalarEnergy::Unload(const SEScalarEnergy& src)
{
  if (!src.IsValid())
    return nullptr;
  cdm::ScalarEnergyData* dst = new cdm::ScalarEnergyData();
  Serialize(src, *dst);
  return dst;
}
void SEScalarEnergy::Serialize(const SEScalarEnergy& src, cdm::ScalarEnergyData& dst)
{
  SEScalarQuantity<EnergyUnit>::Serialize(src, *dst.mutable_scalarenergy());
}